sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs

```diff
@@ -258,17 +193,66 @@ public override void VisitPropertyDeclaration(PropertyDeclarationSyntax node)
+                    // _.node : node._
+                    return node == memberAccess.Name ? ParentAccessType(memberAccess) : SymbolAccess.Read;
+                case MemberBindingExpressionSyntax memberBinding:
+                    // _?.node : node?._
+                    return node == memberBinding.Name ? ParentAccessType(memberBinding) : SymbolAccess.Read;
+                case AssignmentExpressionSyntax assignmentExpression:
+                    // Ignoring distinction assignmentExpression.IsKind(SyntaxKind.SimpleAssignmentExpression) between
+                    // "node = _" and "node += _" both are considered as Write and rely on the parent to know if its read.
+                    //  node = _ : _ = node
+                    return node == assignmentExpression.Left ? SymbolAccess.Write | ParentAccessType(assignmentExpression) : SymbolAccess.Read;
+                case ArgumentSyntax argument:
+                    if (argument.RefOrOutKeyword.IsKind(SyntaxKind.OutKeyword))
+                    {
+                        //  out Type node : out node
+                        return SymbolAccess.Write;
+                    }
+                    else if (argument.RefOrOutKeyword.IsKind(SyntaxKind.RefKeyword))
+                    {
+                        //  ref node
+                        return SymbolAccess.ReadWrite;
+                    }
+                    else
+                    {
+                        return SymbolAccess.Read;
+                    }
+                case ExpressionSyntax expressionSyntax when expressionSyntax.IsAnyKind(IncrementKinds):
+                    // node++
+                    return SymbolAccess.Write | ParentAccessType(expressionSyntax);
+                case ArrowExpressionClauseSyntax arrowExpressionClause when arrowExpressionClause.Parent is MethodDeclarationSyntax arrowMethod:
+                    return arrowMethod.ReturnType != null && arrowMethod.ReturnType.IsKnownType(KnownType.Void, getSemanticModel(arrowMethod))
+                        ? SymbolAccess.None
+                        : SymbolAccess.Read;
+                default:
+                    return SymbolAccess.Read;
+            }
+        }
+
         /// <summary>
         /// Given a node, it tries to get the symbol or the candidate symbols (if the compiler cannot find the symbol,
         /// .e.g when the code cannot compile).
         /// </summary>
         /// <returns>List of symbols</returns>
-        private IEnumerable<ISymbol> GetSymbols<TSyntaxNode>(TSyntaxNode node, Func<TSyntaxNode, bool> condition)
+        private IEnumerable<ISymbol> GetSymbols<TSyntaxNode>(TSyntaxNode node)
             where TSyntaxNode : SyntaxNode
         {
-            return condition(node)
-                ? GetCandidateSymbols(getSemanticModel(node).GetSymbolInfo(node))
-                : Enumerable.Empty<ISymbol>();
+            return GetCandidateSymbols(getSemanticModel(node).GetSymbolInfo(node));
```

Comment:
Local function `GetCandidateSymbols` doesn't have to exist anymore after this update.